buckets/redis/bucket_factory.go

```diff
@@ -164,6 +168,64 @@ func (bf *bucketFactory) reconnectToRedis(oldClient *redis.Client) {
 	}
 }
 
+func (bf *bucketFactory) handleConnectionFailure(oldClient *redis.Client) {
+	bf.Lock()
+	defer bf.Unlock()
+
+	if oldClient == bf.client && !bf.connectionNeedsResolution {
+		logging.Print("Attempting to establish new connection to redis")
+		bf.connectionNeedsResolution = true
+		go bf.establishNewConnectionToRedis(oldClient)
+	}
+}
+
+func (bf *bucketFactory) establishNewConnectionToRedis(oldClient *redis.Client) {
+	client := oldClient
+	numsTried := 0
+	exponentialDelay := 1 * time.Second
+	exponentialDelayCeiling := 5 * time.Minute
+
+	for disconnected := true; disconnected; {
+		attemptsRemaining := bf.connectionRetries
+
+		for attemptsRemaining > 0 {
+			attemptsRemaining--
+			numsTried++
+			bf.reconnectToRedis(client)
+
+			client = bf.Client().(*redis.Client)
+			_, err := client.Ping().Result()
+			if err == nil {
+				disconnected = false
+				break
+			}
+			time.Sleep(1 * time.Second)
```

Comment:
i'd add logging here, that we failed to connect and will retry in 1s

Reply:
this would fill up the log and add noise for other log statements.

Reply:
which is good. I want to see in the logs that process is trying to reconnect

Reply:
Agree with @lukaszx0 - I think that would be incredibly useful to have.